src/rdkafka_sasl_oauthbearer.c

```diff
@@ -0,0 +1,1458 @@
+                        token->token_value = jws;
+                        token->md_lifetime_ms = now_wallclock_millis +
+                                parsed.life_seconds * 1000;
+                        token->md_principal_name = strdup(parsed.principal);
+                        token->extensions = extensionv;
+                        token->extension_size = 2 * extension_pair_count;
+                }
+        }
+        RD_IF_FREE(parsed.principal_claim_name, rd_free);
+        RD_IF_FREE(parsed.principal, rd_free);
+        RD_IF_FREE(parsed.scope_claim_name, rd_free);
+        RD_IF_FREE(parsed.scope_csv_text, rd_free);
+        rd_list_destroy(&parsed.extensions);
+
+        return r;
+}
+
+/**
+ * @brief Default SASL/OAUTHBEARER token refresh callback that generates an
+ * unsecured JWS as per https://tools.ietf.org/html/rfc7515#appendix-A.5.
+ *
+ * This method interprets \c sasl.oauthbearer.config as space-separated
+ * name=value pairs with valid names including principalClaimName,
+ * principal, scopeClaimName, scope, and lifeSeconds. The default
+ * value for principalClaimName is "sub".  The principal must be specified.
+ * The default value for scopeClaimName is "scope", and the default value
+ * for lifeSeconds is 3600.  The scope value is CSV format with the
+ * default value being no/empty scope. For example:
+ * "principalClaimName=azp principal=admin scopeClaimName=roles
+ * scope=role1,role2 lifeSeconds=600".
+ * 
+ * SASL extensions can be communicated to the broker via
+ * extension_<extensionname>=value. For example:
+ * "principal=admin extension_traceId=123".  Extension names and values
+ * must comnform to the required syntax as per
+ * https://tools.ietf.org/html/rfc7628#section-3.1
+ * 
+ * All values -- whether extensions, claim names, or scope elements -- must not
+ * include a quote (") character.  The parsing rules also imply that names
+ * and values cannot include a space character, and scope elements cannot
+ * include a comma (,) character.
+ * 
+ * The existence of any kind of parsing problem -- an unrecognized name,
+ * a quote character in a value, an empty value, etc. -- is a fatal error.
+ * 
+ * Unsecured tokens are not to be used in production -- they are only good for
+ * testing and development purposess -- so while the inflexibility of the
+ * parsing rules is acknowledged, it is assumed that this is not problematic. 
+ */
+void rd_kafka_oauthbearer_unsecured_token(rd_kafka_t *rk, void *opaque) {
+        char errstr[512];
+        struct rd_kafka_sasl_oauthbearer_token token = {0};
+        if (rd_kafka_oauthbearer_unsecured_token_internal(
+            &token,
+            rk->rk_conf.sasl.oauthbearer_config,
+            rd_uclock() / 1000,
+            errstr, sizeof(errstr)) == -1) {
+                rd_kafka_set_fatal_error(rk,
+                        RD_KAFKA_RESP_ERR__INVALID_ARG, "%s", errstr);
+        } else
+                rd_assert(!rd_kafka_oauthbearer_set_token(
```

Comment:
Good point.  Yes, it can fail due to user input.  The code now looks like this:

```
        if (rd_kafka_oauthbearer_unsecured_token_internal(
                &token, rk->rk_conf.sasl.oauthbearer_config, rd_uclock() / 1000,
                errstr, sizeof(errstr))
            || rd_kafka_oauthbearer_set_token(rk, token.token_value,
                token.md_lifetime_ms, token.md_principal_name,
                (const char **)token.extensions, token.extension_size,
                errstr, sizeof(errstr)))
                        rd_kafka_op_err(rk, RD_KAFKA_RESP_ERR__AUTHENTICATION,
                                "%s", errstr);
```